safehttp/plugins/htmlinject/htmlinject.go

```diff
@@ -0,0 +1,193 @@
+// an anti-XSRF token. The rewritten template expects the XSRFToken Func to be available
+// in the template to provide tokens and sets the name for the sent value to be "xsrftoken".
+var XSRFTokensDefault = XSRFTokens(`<input type="hidden" name="xsrf-token" value="{{XSRFToken}}">`)
+
+// XSRFTokens constructs a Config to add the given string as a child node to forms.
+func XSRFTokens(inputTag string) Config {
+	return Config{Rule{
+		Name:     "XSRFTokens on forms",
+		OnTag:    "form",
+		AddNodes: []string{inputTag}}}
+}
+
+// Transform rewrites the given template according to the given configs.
+// If the passed io.Rewriter has a `Size() int64` method it will be used to pre-allocate buffers.
+func Transform(src io.Reader, cfg ...Config) (tpl string, _ error) {
+	rw := rewriter{
+		rules:     map[string][]Rule{},
+		tokenizer: html.NewTokenizer(src),
+		out:       &strings.Builder{},
+	}
+	if sizer, ok := src.(interface{ Size() int64 }); ok {
+		rw.out.Grow(int(sizer.Size()))
+	}
+	for _, c := range cfg {
+		for _, r := range c {
+			rw.rules[r.OnTag] = append(rw.rules[r.OnTag], r)
+		}
+	}
+	if err := rw.rewrite(); err != nil {
+		return "", fmt.Errorf("transforming template: %v", err)
+	}
+	return rw.out.String(), nil
+}
+
+type rewriter struct {
+	// tag -> rules for that tag
+	rules     map[string][]Rule
+	tokenizer *html.Tokenizer
+	out       *strings.Builder
+}
+
+// emitRaw copies the current raw token to the output.
+func (r rewriter) emitRaw() error {
+	_, err := r.out.Write(r.tokenizer.Raw())
+	return err
+}
+
+// rewrite runs the rewriter.
+func (r rewriter) rewrite() error {
+	for {
+		switch tkn := r.tokenizer.Next(); tkn {
+		case html.ErrorToken:
+			if err := r.tokenizer.Err(); !errors.Is(err, io.EOF) {
+				return err
+			}
+			// We got EOF, let's just emit the last token and exit.
+			return r.emitRaw()
+		case html.StartTagToken, html.SelfClosingTagToken:
+			if err := r.processTag(); err != nil {
+				return err
+			}
+		default:
+			if err := r.emitRaw(); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (r rewriter) processTag() error {
+	// Copy raw tokens to better formats
+	var (
+		tagname    string
+		attributes = map[string]string{}
+		raw        = make([]byte, len(r.tokenizer.Raw()))
+	)
+	{
+		copy(raw, r.tokenizer.Raw())
+		tag, hasAttr := r.tokenizer.TagName()
+		tagname = string(tag)
+		for hasAttr {
+			key, val, more := r.tokenizer.TagAttr()
+			hasAttr = more
+			attributes[string(key)] = string(val)
+		}
+	}
+
+	// Filter rules by attributes
+	var triggeredRules []Rule
+	{
+		rules := r.rules[tagname]
+		for _, r := range rules {
+			match := true
+			for k, v := range r.WithAttributes {
+				if attributes[k] != v {
+					match = false
+				}
+			}
+			if match {
+				triggeredRules = append(triggeredRules, r)
+			}
+		}
+	}
+
+	// Emit the rewritten HTML
+	{
+		// Write the "<" symbol and the tag name, e.g. "<script"
```

Comment:
I am slightly concerned about the possibility that this may go wrong, e.g. if the input string has whitespace between the `<` and the tag name (is that allowed?). Maybe use `len(tagname) + bytes.Index(raw, tagname)`?

Reply:
that is not valid HTML and would be rejected by any html template. Also, that is not a token that is emitted by the tokenizer.